src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -1656,21 +1750,33 @@ public ProcessOutputHandler(Process process, BlockingCollection<ProcessOutputObj
             if (process.StartInfo.RedirectStandardError) { _refCount++; }
 
             // once we have _refCount, we can start processing
-            if (process.StartInfo.RedirectStandardOutput)
-            {
-                _isFirstOutput = true;
-                _isXmlCliOutput = false;
-                process.OutputDataReceived += OutputHandler;
-                process.BeginOutputReadLine();
-            }
-
             if (process.StartInfo.RedirectStandardError)
             {
                 _isFirstError = true;
                 _isXmlCliError = false;
                 process.ErrorDataReceived += ErrorHandler;
                 process.BeginErrorReadLine();
             }
+
+            stdOutDrainer = null;
+
+            if (!process.StartInfo.RedirectStandardOutput)
+            {
+                return;
+            }
+
+            if (!ExperimentalFeature.IsEnabled(ExperimentalFeature.PSNativeCommandPreserveBytePipe)
+                || stdOutDestination is null)
+            {
+                _isFirstOutput = true;
+                _isXmlCliOutput = false;
+                process.OutputDataReceived += OutputHandler;
+                process.BeginOutputReadLine();
+                return;
+            }
+
+            stdOutDrainer = _stdOutDrainer = stdOutDestination.Bind(stdOutSource);
+            stdOutDrainer.BeginReadChunks();
```

Comment:
`stdOutDrainer.BeginReadChunks()` is called in `InitNativeProcess` which happens in `Prepare`. Doesn't this mean that the downstream native process's input stream needs to be already opened when the current native command processor is still running `Prepare`?

As of today, the downstream native command's `_nativeProcess` is not yet created when the `BeginReadChunks()` from the upstream native command starts to run, so there seems to be a race condition: `_callback` gets called within `BeginReadChunks()`, which calls `_nativeProcess.StandardInput.BaseStream` on the downstream native command processor while `_nativeProcess` is still `null`.

Reply:
> `stdOutDrainer.BeginReadChunks()` is called in `InitNativeProcess` which happens in `Prepare`. Doesn't this mean that the downstream native process's input stream needs to be already opened when the current native command processor is still running `Prepare`?
> 
> As of today, the downstream native command's `_nativeProcess` is not yet created when the `BeginReadChunks()` from the upstream native command starts to run, so there seems to be a race condition: `_callback` gets called within `BeginReadChunks()`, which calls `_nativeProcess.StandardInput.BaseStream` on the downstream native command processor while `_nativeProcess` is still `null`.

You're absolutely correct. Initially I was forcing the upstream commands `Prepare` method to be invoked first but while cleaning up I really wanted to not change the order there.

Unfortunately a minimal hello world rust app is enough to hit a NRE here, so great catch! I'm thinking a sync primitive to block until ready will still be more clean than trying to force the upstream to be prepared first. Working on that now

Reply:
Alright I've added some synchronization. I'm less convinced that it's cleaner than forcing downstream commands to `Prepare` first, let me know what you think. I'm still stress testing but it's looking good so far.